src/main/java/blusunrize/immersiveengineering/common/data/Recipes.java

```diff
@@ -107,77 +112,581 @@ protected void registerRecipes(@Nonnull Consumer<IFinishedRecipe> out)
+		SingleItemRecipeBuilder.func_218648_a(Ingredient.fromItems(StoneDecoration.concreteTile), IEBlocks.toSlab.get(StoneDecoration.concreteTile))
+			.func_218643_a("has_concrete", this.hasItem(StoneDecoration.concrete))
+			.func_218647_a(out, toRL("concrete_tile_slab_from_concrete_tile_stonecutting"));
+		SingleItemRecipeBuilder.func_218648_a(Ingredient.fromItems(StoneDecoration.concreteTile), StoneDecoration.concreteStairs[1])
+			.func_218643_a("has_concrete", this.hasItem(StoneDecoration.concrete))
+			.func_218647_a(out, toRL("concrete_tile_stairs_from_concrete_tile_stonecutting"));
+		SingleItemRecipeBuilder.func_218648_a(Ingredient.fromItems(StoneDecoration.concreteLeaded), IEBlocks.toSlab.get(StoneDecoration.concreteLeaded))
+			.func_218643_a("has_concrete", this.hasItem(StoneDecoration.concrete))
+			.func_218647_a(out, toRL("concrete_leaded_slab_from_concrete_leaded_stonecutting"));
+		SingleItemRecipeBuilder.func_218648_a(Ingredient.fromItems(StoneDecoration.concreteLeaded), StoneDecoration.concreteStairs[2])
+			.func_218643_a("has_concrete", this.hasItem(StoneDecoration.concrete))
+			.func_218647_a(out, toRL("concrete_leaded_stairs_from_concrete_leaded_stonecutting"));
 
+		ShapedRecipeBuilder.shapedRecipe(WoodenDecoration.treatedScaffolding, 6)
+			.patternLine("iii")
+			.patternLine(" s ")
+			.patternLine("s s")
+			.key('i', IETags.getItemTag(IETags.treatedWood))
+			.key('s', IETags.treatedStick)
+			.addCriterion("has_treated_planks", hasItem(IETags.getItemTag(IETags.treatedWood)))
+			.addCriterion("has_treated_sticks", hasItem(IETags.treatedStick))
+			.build(out);
+		ShapedRecipeBuilder.shapedRecipe(MetalDecoration.aluScaffolding.get(MetalScaffoldingType.STANDARD), 6)
+			.patternLine("iii")
+			.patternLine(" s ")
+			.patternLine("s s")
+			.key('i', IETags.getTagsFor(EnumMetals.ALUMINUM).ingot)
+			.key('s', IETags.aluminumRod)
+			.addCriterion("has_alu_ingot", hasItem(IETags.getTagsFor(EnumMetals.ALUMINUM).ingot))
+			.addCriterion("has_alu_sticks", hasItem(IETags.aluminumRod))
+			.build(out);
+		ShapedRecipeBuilder.shapedRecipe(MetalDecoration.steelScaffolding.get(MetalScaffoldingType.STANDARD), 6)
+			.patternLine("iii")
+			.patternLine(" s ")
+			.patternLine("s s")
+			.key('i', IETags.getTagsFor(EnumMetals.STEEL).ingot)
+			.key('s', IETags.steelRod)
+			.addCriterion("has_steel_ingot", hasItem(IETags.getTagsFor(EnumMetals.STEEL).ingot))
+			.addCriterion("has_steel_sticks", hasItem(IETags.steelRod))
+			.build(out);
+		for(int i=0;i<MetalScaffoldingType.values().length;i++)
+		{
+			MetalScaffoldingType from = MetalScaffoldingType.values()[i];
+			MetalScaffoldingType to = null;
+			if((i+1)>=MetalScaffoldingType.values().length) // Meant to prevent this whole thing to throw IndexOutOfBounds.
+			{
+				to = MetalScaffoldingType.STANDARD; // And to have wood_top -> standard conversion
+			}
+			else
+			{
+				to = MetalScaffoldingType.values()[i+1];
+			}
```

Comment:
What I suggested in my comment shouldn't throw AIOOBEs, that's what the `%numScaffoldingTypes` is for. If you want to keep things this way you should remove the `{}`s for the if, in IE we don't use them for statements with a single-line body (I don't really agree with that style in general, but this feels like far too "big" vertically).